### ISSUE TITLE
Matcher-based Listener filters

### ISSUE THREAD

**Issue by howardjohn** [envoyproxy/envoy#35958]
*Title*: Matcher-based Listener filters

*Description*:
Currently, we use tls_inspector and http_inspector in a variety of places to inspect traffic. The problem with these is its not safe to unconditionally enable them, as some protocols are "server first" protocols where the server sends the first bytes of traffic.

In the past, we worked around this by using the `FilterDisabled` field on the listener filter. For our simple use case we had at the time, this was sufficient because we only needed to classify based on port. We used many bind_to_port=false listeners, rather than 1 large listener with a match on IP, effectively giving us an IP-match implicitly 

We are moving towards a more complex matching model, reliant on a single listener with lots of chains and using the filter_chain_matcher API. This breaks our ability to disable the filter appropriately.

## Option 1: integrated into the matcher
In the ideal scenario, the listener filters could be enable in the matching tree itself:
```yaml
filterChainMatcher:
  matcherTree:
    input:
      name: ip
      rangeMatchers:
      - ranges:
        - addressPrefix: 240.240.0.2
        onMatch:
          listenerFilters: ["http_inspector"] # THIS WOULD BE NEW
          exactMatchMap:
            input:
              name: port
            map:
              "465":
                action: cluster-465
              "587":
                action: cluster-587
```

Basically we do some preliminary matching, then run the listener filter, then can do more matching. Obviously the user will need to ensure they run the filter before matching the output of it.

## Option 2: a new dedicated match

Rather than the `FilterDisabled` approach which just handles ports, we could have a `FilterDisabledMatcher` or equivalent which uses the new matcher framework.

The downside of this approach is that it requires generating a mirror matcher tree. This may be expensive and complex to maintain

cc @keithmattix

**Comment by keithmattix** [comment]
I'm not sure option 1 is feasible unfortunately. If you have a FCM based off of SNI, how could it match without running tls inspector first?

**Comment by howardjohn** [comment]
You would need to put the "use tls inspector" before the SNI match

**Comment by keithmattix** [comment]
Am I understanding you correctly that we would need to also have a (small) dependency tree of shorts for matches and listener filters in order to know if there's a listener filter that needs to be inserted even if the control plane/user didn't specify it? In other words: if I specify a server_name match without putting tls_inspector in my listenerFilter list, will that filter get added or will the match always fail?

**Comment by howardjohn** [comment]
Maybe this is more clear:
```yaml
matcher:
  # Match IP first
  matchType: IP
  matchers:
    1.2.3.4:
      onMatch:
        # We match, run inspector
        enableListenerFilter: my-tls_inspector
        matchTree:
          # Now match SNI
          matchType: SNI
          ...
    3.4.5.6:
      onMatch: # Match SNI without enabling the listener... this is broken, will never match; users should just not do this
        matchTree:
          # Now match SNI
          matchType: SNI
          ...
    5.6.7.8:
      onMatch: # here I don't enable tls inspector but also do not needed
        matchTree:
          # Now match port
          matchType: port
          ...
listener_filters:
- name: my-tls_inspector
  typed_config: ...
  use_matcher: true # signal to not always enable. But we don't want to put the full spec in the tree probably ,just  the named reference
- name: my-http_inspector
  typed_config: ...
  use_matcher: true

```

**Comment by keithmattix** [comment]
Ok got it thanks. So basically this would add the ability to defer listener filter execution until after matching. This would make the downstream_listener_pre_cx metric really strange...but otherwise might be feasible

**Comment by howardjohn** [comment]
Its not really _after_ matching, its between matching. I match IP, then run the tls_inspector,then can match on SNI.

**Comment by phlax** [comment]
cc @ggreenway

**Comment by ggreenway** [comment]
That's a really interesting idea. I think one big complication is that evaluating the matcher tree will have to be able to stop in the middle and resume later. But I'm interested in seeing where investigation into this goes.

**Comment by keithmattix** [comment]
I need to get up to speed on the matching logic, but I can spearhead the investigation with my team at least. /cc @Stevenjin8 @krinkinmu

**Comment by kyessenov** [comment]
Technically the unified matcher can be extended in this lazy fashion, if we give the ability to inputs to start/pause listener filter executions. Essentially, instead of the current SNI input, we'd add a "lazy SNI input" that activates the TLS inspector till completion. This is also somewhat related to the "nested transport sockets" issue since currently, we have no ability to inspect HTTP within TLS, so the inspector don't really combine.

**Comment by keithmattix** [comment]
Not stale

**Comment by krinkinmu** [comment]
@howardjohn I'm working on this issue now and wanted to make sure that I undertand the problem correctly.

My understanding currently is that with migration to a single listener, Istio cannot enable/disable TLS inspector based on the IP address, because filter_disabled field only allows to set conditions on ports and not IP addresses. Is that correct?

If my understanding is correct, then extending filter_disabled to allow conditions based on IP addresses (in addition to ports) would be the most straighforward way to address the problem. And it's my understanding that the option 2 is suggesting exactly that.

However, I don't think I fully understand the downside of that option that you pointed out. Can you elaborate briefly on what kind of config you had in mind for the option 2 and what are biggest pain points in generating/mainting it?

**Comment by krinkinmu** [comment]
Sorry for the dealy, I was a bit distructed by some other work. I started working on this and trying to add support for pausing filter chain matching to allow inputs to pause and run listener filters on deman. I don't have a complete implementation yet, but I think I have something that makes sense to share and get some early feedback.

I posted in the envoy-dev slack chat and want to post it here as well, this is the change that makes sense to discuss: https://github.com/krinkinmu/envoy/commit/9258388cca60ba98091ca904c205f3f3e8ff62d0. In this change I'm trying to change the FilterChainManager API to take a callback instead of returning a matching filter chain directly. That is to allow findFilterChain to pause if we need to run a listener filter.

The change is getting somewhat large, that's why I'm interested in some early feedback on the approach (and will keep working on other parts of the implementation in the meantime).

**Comment by zohebk8s** [comment]
@howardjohn @keithmattix Is Sybase a server-first protocol? How can we determine or conclude that? It's not listed here - https://istio.io/latest/docs/ops/deployment/application-requirements/

I'm troubleshooting a connectivity issue between a Spring Boot application and a Sybase database running outside the Istio service mesh. When the Istio sidecar is injected, the connection hangs. But if I remove the sidecar from the same pod, the connection succeeds.

This leads me to suspect that Sybase uses a server-first protocol, which conflicts with Istio’s protocol sniffing behavior.

From the Istio codebase, I understand that naming the port with a tcp- prefix (e.g., tcp-sybase) disables the HTTP inspector and builds a TCP-only filter chain, which should allow server-first protocols to work.

Can you confirm:

Is Sybase officially considered a server-first protocol in Istio’s listener logic?

Is naming the port tcp-sybase in the Service manifest mandatory to avoid protocol sniffing for server-first flows?

apiVersion: v1
kind: Service
metadata:
  name: sybase-service
spec:
  type: ClusterIP
  clusterIP: 10.x.x.x
  ports:
    - name: tcp-sybase  # Required for server-first?
      protocol: TCP
      port: 50001
      targetPort: 50001
